### PR TITLE
HPCC-14452 CSocketEpollThread - no exception for bad file descriptor

### PR DESCRIPTION
**Pull request by mckellyln**
@jakesmith please review.

Signed-off-by: Mark Kelly mark.kelly@lexisnexis.com

### CODE REVIEW

File: system/mp/mpcomm.cpp [hunk 1]
```diff
@@ -729,7 +729,8 @@ protected: friend class CMPPacketReader;
             try {
                 StringBuffer str;
 #ifdef _TRACE
-                LOG(MCdebugInfo(100), unknownJob, "MP: connecting to %s",remoteep.getUrlStr(str).toCharArray());
+                str.clear();
```
Comment (jakesmith) [inline system/mp/mpcomm.cpp]: str will be empty, so this line is not necessary.

File: system/mp/mpcomm.cpp [hunk 2]
```diff
@@ -831,13 +832,23 @@ protected: friend class CMPPacketReader;
 #ifdef _TRACE
                                 EXCLOG(e, "MP: Failed to connect");
 #endif
+                                // if other side closes because of stale connection on same host:port, connect again
+                                if (e->errorCode() == JSOCKERR_graceful_close)
```
Comment (jakesmith) [inline system/mp/mpcomm.cpp]: This change makes sense afaics, but don't see how related to epoll, so should be separate commit I think.

File: system/mp/mpcomm.cpp [hunk 3]
```diff
@@ -831,13 +832,23 @@ protected: friend class CMPPacketReader;
...
+                                {
+                                    LOG(MCdebugInfo(100), unknownJob, "MP: Retrying (other side closed connection, probably due to clash)");
+                                    e->Release();
+                                    break;
```
Comment (jakesmith) [inline system/mp/mpcomm.cpp]: should this check happen after retrycount check/abort below?

i.e. should a graceful close be considered 1 of the retry attempts.

**Comment by jakesmith** [comment]
@mckellyln - I think this should be split into 2 commits

File: system/mp/mpcomm.cpp (hunk 3 above)
Comment (mckellyln) [inline system/mp/mpcomm.cpp]: I will put this into a new JIRA issue.  I can look at where it should go and take retrycount/timeout into account.

**Comment by mckellyln** [comment]
Code updated based on review.  Please re-review @jakesmith when you get a chance.

File: system/jlib/jsocket.cpp [hunk 4]
```diff
@@ -4306,8 +4307,7 @@ class CSocketEpollThread: public CSocketBaseThread
         // if another thread closed fd before here don't fail
```
Comment (jakesmith) [inline system/jlib/jsocket.cpp]: v.minor: >don't fail >> don't issue warning

**Comment by jakesmith** [comment]
@mckellyln - 1 v.minor comment, looks good to merge.
Is there another JIRA yet for the graceful close change for reference?

**Comment by mckellyln** [comment]
Yes, its hpcc-14536
